clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -2801,26 +2801,57 @@ void CodeGenFunction::EmitStoreThroughExtVectorComponentLValue(RValue Src,
                                                                LValue Dst) {
   llvm::Value *SrcVal = Src.getScalarVal();
   Address DstAddr = Dst.getExtVectorAddress();
+  const llvm::Constant *Elts = Dst.getExtVectorElts();
   if (DstAddr.getElementType()->getScalarSizeInBits() >
       SrcVal->getType()->getScalarSizeInBits())
     SrcVal = Builder.CreateZExt(
         SrcVal, convertTypeForLoadStore(Dst.getType(), SrcVal->getType()));
 
-  // HLSL allows storing to scalar values through ExtVector component LValues.
-  // To support this we need to handle the case where the destination address is
-  // a scalar.
-  if (!DstAddr.getElementType()->isVectorTy()) {
-    assert(!Dst.getType()->isVectorType() &&
-           "this should only occur for non-vector l-values");
-    Builder.CreateStore(SrcVal, DstAddr, Dst.isVolatileQualified());
+  if (getLangOpts().HLSL) {
+    llvm::Type *DestAddrTy = DstAddr.getElementType();
+    // HLSL allows storing to scalar values through ExtVector component LValues.
+    // To support this we need to handle the case where the destination address
+    // is a scalar.
+    if (!DestAddrTy->isVectorTy()) {
+      assert(!Dst.getType()->isVectorType() &&
+             "this should only occur for non-vector l-values");
+      Builder.CreateStore(SrcVal, DstAddr, Dst.isVolatileQualified());
+      return;
+    }
+
+    // In HLSL, storing to individual elements of a vector through ExtVector
+    // components needs to be handled as separate store instructions. We need to
+    // avoid the load/modify/store sequence to prevent overwriting other
+    // elements that might be getting updated in parallel.
+    // If we are updating multiple elements, Dst and Src are vectors; for
+    // a single element update they are scalars.
+    const VectorType *VTy = Dst.getType()->getAs<VectorType>();
+    unsigned NumSrcElts = VTy ? VTy->getNumElements() : 1;
```

Comment:
doesn't the above if case handle when Dst.getType isn't a vector? Could we assert that VTy != null instead?

Reply:
No, there are actually three cases:

```
  int i;
  int4 vec;

  i.x = 1.0; // case 1
  vec.y = 1; // case 2
  vec.xz = { 3, 4}; // case 3
```

Case 1 is storing a scalar to scalar via vector swizzle and it is handled by the code above this comment.
Case 2 is storing a scalar to vector via single element swizzle and `VTy` is going to be null in this case.
Case 3 is storing multiple scalars to a vector via multi-element swizzle and `VTy != nullptr`.